### DISCUSSION TITLE
Creating an Egsphant file in python gives an error for plane positions

### DISCUSSION THREAD

**Discussion by ShahinGHD** [nrc-cnrc/EGSnrc#983] [Questions] [answered]
Hi everyone,
I am generating the egsphant file using a code I wrote in python.
when I execute the EGS simulation it says:

```
EGS_PlanesT::EGS_PlanesT: plane positions must be in increasing order.
```

I am 100 percent sure that my plane positions are in increasing order since I use np.sort in python to sort them. When I generate the same egsphant file in DOSXYZnrc GUI and replace the plane positions to my file, the problem is solved and simulations runs. It shows that my egsphant generated in python works well except for the planes.

Why am I getting this error although my coordinates are in increasing order and I have no duplicate number?

thanks

**Comment by rtownson** [comment]
What's the difference between the DOSXYZnrc GUI planes, and yours? Copy/paste them here? Also keep in mind that the voxel size must be constant along each axis.

**Comment by ShahinGHD** [comment]
Thanks for the comment,
The weird thing is I get to make different voxel sizes with the GUI and the simulation is fine; the geometry is that I have two cubic boxes inside each other. The bigger one has 1cm voxel size, and the inner one has a 0.01 cm voxel size
I am getting a 3ddose file from the simulation and somehow, it is fine with different voxel sizes.

How is that possible? Because I know I should have equal voxel sizes.

the egsphant from the GUI is:

```
-5.5 -4.5 -3.5 -2.5 -1.5 -0.5 -0.49 -0.48 -0.47 -0.46 -0.45 -0.44 -0.43 -0.42 -0.41 -0.40 -0.39 -0.38
-0.37 -0.36 -0.35 -0.34 -0.33 -0.32 -0.31 -0.30 -0.29 -0.28 -0.27 -0.26 -0.25 -0.24 -0.23 -0.22 -0.21
-0.20 -0.19 -0.18 -0.17 -0.16 -0.15 -0.14 -0.13 -0.12 -0.11 -0.10 -0.09 -0.08 -0.07 -0.06 -0.05 -0.04
-0.03 -0.02 -0.01 0.00 0.01 0.02 0.03 0.04 0.05 0.06 0.07 0.08 0.09 0.10 0.11 0.12 0.13 0.14 0.15 
.16 0.17 0.18 0.19 0.20 0.21 0.22 0.23 0.24 0.25 0.26 0.27 0.28 0.29 0.30 0.31 0.32 0.33 0.34 0.35
0.36 0.37 0.38 0.39 0.40 0.41 0.42 0.43 0.44 0.45 0.46 0.47 0.48 0.49 0.50 0.51 0.52 0.53 0.54 0.55
1.55 2.55 3.55 4.55 5.55
```

you see there is different sizes

my python code generates this:

```
-5.50 -4.50 -3.50 -2.50 -1.50 -0.53 -0.51 -0.50 -0.50 -0.49 -0.48 -0.47 -0.46 -0.45 -0.44 -0.43 -0.42
-0.41 -0.40 -0.39 -0.38 -0.37 -0.36 -0.35 -0.34 -0.33 -0.32 -0.31 -0.30 -0.29 -0.28 -0.27 -0.26 -0.25
-0.24 -0.23 -0.22 -0.21 -0.20 -0.19 -0.18 -0.17 -0.16 -0.15 -0.14 -0.13 -0.12 -0.11 -0.10 -0.09 -0.08
-0.07 -0.06 -0.05 -0.04 -0.03 -0.02 -0.01 0.00 0.01 0.02 0.03 0.04 0.05 0.06 0.07 0.08 0.09 0.10 0.11
0.12 0.13 0.14 0.15 0.16 0.17 0.18 0.19 0.20 0.21 0.22 0.23 0.24 0.25 0.26 0.27 0.28 0.29 0.30 0.31
0.32 0.33 0.34 0.35 0.36 0.37 0.38 0.39 0.40 0.41 0.42 0.43 0.44 0.45 0.46 0.47 0.48 0.49 0.50 0.51
0.53 1.50 2.50 3.50 4.50 5.50
```

they are equal except the last 5 numbers,
the most weird thing is when I correct them by hand to be exactly similar, I still get the error above.

**Reply by rtownson** [reply]
It looks like the varying voxel sizes might indeed be ok, at least from the code I've looked at so far it does indeed seem to use the correct volume for the dose calculation.

The error you're getting is strange. Does it still happen if you copy/paste the whole set of boundaries? I wonder if the 'newline' characters are a problem. It's tricky for me to know without the phantoms in-hand. If you are able to, you could share them with a google drive link.

**Reply by mchamberland** [reply, marked as answer]
There is at least one duplicate bound in the Python version: there are two -0.50. That would likely cause an issue.

**Reply by ShahinGHD** [reply]
Thanks for the reply, yes, that was the problem. I am working on a way to eliminate this duplicate number.

But can anyone confirm that egsphants or 3ddose files actually WORK with different voxel sizes? Because mine is working.

I am not sure how the egs_mird (based on egs++) program that I am using is addressing regions when I have two phantoms, one inside the other one with different voxel sizes, but it is actually working

**Reply by rtownson** [reply]
If you have two XYZ geometries that are unioned together, how are you generating a single 3ddose file for them? The combined geometry would be a union so would not be accepted by egs_dose_scoring, which only accepts an XYZ geometry.

**Reply by ShahinGHD** [reply]
When I use the egsphant file, I define those two geometries, one inside the other with different voxel sizes. The density inside these two boxes is also different. 
gunion library is not used here. It's egs_glib library I am using.
The egsphant is either come from DOSXYZ_GUI or from the code I wrote in Python.
The output is 3ddose file considering all the voxels with different sizes.


I did use the union library too, it is possible to define two boxes with two different voxel sizes and combine them with gunion.
The limitation here is that in the "scoring" block in egsinp file, you have to set the geometry for either of the boxes, but not for the union one.
This means you can score the dose in one of the m and not both.

I hope this answers your question.

btw: if you know any way to combine two egs_xyzgeometrys with different voxel sizes and score the dose in both of them please let me know.
Thanks

**Reply by rtownson** [reply]
If you have generated a single egsphant file using dosxyz_gui and it visualizes correctly when you open the egsinp file in egs_view, then I imagine everything is fine. But I can't see how it's possible to have two completely different resolutions, with a single egsphant file. At best you would have a box with higher res in the middle, but also high res "stripes" through the larger box out from the 6 surfaces of the smaller box (so the larger box would not entirely have a larger voxel size, really just in the corners of the cube). Is that what you mean? If so, yes that should work fine.

**Reply by ShahinGHD** [reply]
I see what you mean. The egs_view has a huge problem opening the egsphant file. 
It takes forever because I have so many voxels. So I am not able to see that, but from the dose scores I get from a single 3ddose file, I would say that the description you gave is not the case.
the inner box res is 0.01x0.01x0.01 cm cubic,(each vox size)
the total size of the inner box is 1x1x1 cm
the voxel sizes of the outer box is also 1x1x1 cm, I am trying to open the egsphant in egs_iew but no luck so far.

**Reply by mchamberland** [reply]
I don't have time to check right now, but I don't think egs_mird uses egs_dose_scoring to output 3ddose. The 3ddose output is probably built into the code for efficiency or some other design reasons.
